docs/NamespaceIsolation.md

```diff
@@ -0,0 +1,58 @@
+# Namespace isolation for BPF objects
+This document describes how eBPF for Windows implements namespace isolation.
+
+## Overview
+BPF objects are by default system wide entities that can be opened via several libbpf APIs (meaning the caller receives
+a fd to the object). While this is useful for the general case, there are some scenarios where BPF applications need to
+be able to isolate themselves from other BPF applications to ensure that their state remains consistent.
+
+To achieve this a BPF application can switch to a designated namespace. Namespaces are identified by a GUID, with the
+default namespace having the zero GUID (all fields in the GUID are zero). All libbpf APIs that return fds are relative
+to the namespace that the application has specified.
+
+## Design details
+Each BPF object will have a namespace identifier associated with it. Each new object created is marked with the current
+namespace identifier. All calls to the execution context to open objects are relative to the current namespace.
+
+### Namespace in ebpf_core_object_t
+All user accessible BPF objects embed the ebpf_core_object_t structure:
+
+```c
+typedef struct _ebpf_core_object
+{
+    ebpf_base_object_t base;              ///< Base object for all reference counted eBPF objects.
+    ebpf_object_type_t type;              ///< Type of this object.
+    ebpf_free_object_t free_function;     ///< Function to free this object.
+    ebpf_zero_ref_count_t zero_ref_count; ///< Function to notify the object that the reference count has reached
+                                            ///< zero.
+    ebpf_object_get_program_type_t get_program_type; ///< Function to get the program type of this object.
+    ebpf_id_t id;                                    ///< ID of this object.
+    ebpf_list_entry_t object_list_entry;             ///< Entry in the object list.
+    volatile int32_t pinned_path_count;              ///< Number of pinned paths for this object.
+    struct _ebpf_epoch_work_item* free_work_item;    ///< Work item to free this object when the epoch ends.
```

Comment:
"free_work_item" sounds like either:
a) a function one can call to free the work item (if "free" is a verb), or
b) a pointer to a work item where the work item itself is free (if "free" is an adjective).

Perhaps "free_object_work_item"?

Reply:
Filed https://github.com/microsoft/ebpf-for-windows/issues/4425 to track this.